content/en/blog/_posts/2023-04-06-seccomp-profiles-edge.md

```diff
@@ -0,0 +1,280 @@
+(YAML) based seccomp profile and then apply it to the process.
+
+The Security Profiles Operator supports a [log enricher feature][enricher],
+which provides additional seccomp related information by parsing the audit logs.
+`spoc run` uses the enricher in the same way to provide more data to the end
+users when it comes to debugging seccomp profiles.
+
+[enricher]: https://github.com/kubernetes-sigs/security-profiles-operator/blob/35ebdda/installation-usage.md#using-the-log-enricher
+
+Now I have to modify the profile to see anything valuable in the output. For
+example, I could remove the allowed `uname` syscall:
+
+```console
+> jq 'del(.syscalls[0].names[] | select(. == "uname"))' /tmp/profile.json > /tmp/no-uname-profile.json
+```
+
+And then try to run it again with the new profile `/tmp/no-uname-profile.json`:
+
+```
+> sudo ./spoc run -p /tmp/no-uname-profile.json ./main
+10:39:12.707798 Reading file /tmp/no-uname-profile.json
+10:39:12.707892 Setting up seccomp
+10:39:12.707920 Load seccomp profile
+10:39:12.707982 Starting audit log enricher
+10:39:12.707998 Enricher reading from file /var/log/audit/audit.log
+10:39:12.709164 Running command with PID: 480512
+panic: operation not permitted
+
+goroutine 1 [running]:
+main.main()
+        /path/to/main.go:10 +0x85
+10:39:12.713035 Unable to run: launch runner: wait for command: exit status 2
+```
+
+Alright, that was expected! The applied seccomp profile blocks the `uname`
+syscall, which results in an "operation not permitted" error. This error is
+pretty generous and does not provide any hint that it got blocked by seccomp.
+It is generally extremely difficult to predict how applications behave if single
+syscalls are forbidden by seccomp. It could be possible that the application
+terminates like in our simple demo, but it could also lead to a strange
+misbehavior and the application does not stop at all.
+
+If I now change the default seccomp action of the profile from `SCMP_ACT_ERRNO`
+to `SCMP_ACT_LOG` like this:
+
+```console
+> jq '.defaultAction = "SCMP_ACT_LOG"' /tmp/no-uname-profile.json > /tmp/no-uname-profile-log.json
+```
+
+Then the log enricher will give us a hint that the `uname` syscall got blocked
+when using `spoc run`:
+
+```
+> sudo ./spoc run -p /tmp/no-uname-profile-log.json ./main
+10:48:07.470126 Reading file /tmp/no-uname-profile-log.json
+10:48:07.470234 Setting up seccomp
+10:48:07.470245 Load seccomp profile
+10:48:07.470302 Starting audit log enricher
+10:48:07.470339 Enricher reading from file /var/log/audit/audit.log
+10:48:07.470889 Running command with PID: 522268
+10:48:07.472007 Seccomp: uname (63)
+```
+
+The application will not terminate any more, but seccomp will log the behavior
+to `/var/log/audit/audit.log` and `spoc` will parse the data to correlate it
+directly to our program. Generating the log messages to the audit subsystem
+comes with a large performance overhead and should be handled with care in
+production systems.
```

Comment:
It may be worth calling out the security risks of running untrusted apps in audit mode in production environments.

Reply:
Yep, mentioned that as well.